### PR TITLE
feat: detached configuration management approach

### PR DESCRIPTION
**Pull request by PeterBaker0**
# Provena PR (major): detached configuration management approach

## Checklist

-   [x] If tests are required for this change, are they implemented?
-   [x] Are user documentation changes required, if so, is there a task to track it and/or is it completed?
-   [ ] If migrations are required, is the process documented below?
-   [ ] If developer/system documentation updates are required, is there a task to track it and/or is it completed?
-   [ ] At least one developer has reviewed this change (unless PR is being used to mark a commit point without need for review)?
-   [x] If change requires update to the [manual testing runsheet](https://confluence.csiro.au/pages/viewpage.action?pageId=1701138616), is there a task to track it and/or is it completed?

## Description

Major overhaul of how configuration is managed. At a high level, the changes include

- removing the need for users to commit CDK and associated configuration directly to the repo (meaning they need a private fork)
- moving configuration into JSON files which are validated against updated pydantic schemas at runtime
- JSON files are .gitignored and helper scripts are introduced into both the local development and pipeline workflows which retrieve these files from a private configuration repository
- introduces a controlled mechanism for retrieving configs
  - introduces `namespace`:= a collection of related deployment configs e.g. an account/organisation
  - introduces `stage`:= the particular deployment to target
  - introduces `base`:= a stage within a namespace which is always included as a base set of private config files to copy
- config bashly script at top level which interacts with the above structured config repo to pull files from the targeted stage using the high level syntax `./config namespace stage` - uses tmp directories to git clone then cleanup - relies on local git permissions but has an override for providing an exiting github repo which is already cloned for use in pipelines
- config methodology = merge the structure at a given namespace/stage into the current provena environment. Gitignore ignores all files in targeted config locations except sample files provided
- brings feature branch into public repository to provide dev friendly workflow for deploying temporary deployments to private AWS accounts in public context, alleviating the need to develop in a private fork for shared features
- moves the github bootstrap and feature branch manager logic into public repo and separate CDK projects which also use config logic
- implements variable substitution JSON syntax with custom implementation to enable env variable substitutions which are flexible enough to handle straight replacements, defaults, conditional with if/else. Special rules apply for string substitution to enable complete control as to how output is replaced without violating JSON syntax requirements (to enable linting and auto formatting)
- created template configuration repo https://github.com/provena/provena-config-repo-template

## Migrations Required

The work to define a well managed migration process for this is greater than handing on an individual basis. 

The high level task will be to export/dump your current configuration to JSON, then validate it (which can be done with any `cdk`  task). Some of the new stacks will require new configs to be setup, but these are straight forward and will be documented.

## Deferred TODOs (all must have ticket)

- Overhaul configuration related documentation
  - infrastructure readme
  - json config documentation
  - ui only / bootstrap config setups
  - ui theming approach 
- replace custom deployment template with rich sample.json and documentation
- overhaul of provena from scratch confluence doc
- token refresh operational procedure update

## Notes for reviewer

Discussion to go through changes and check things work as expected.

### CODE REVIEW

**Comment by PeterBaker0** [comment]
FYI - have updated readme at configuration repo for RRAP - see our github org and private config repo

File: featb/src/initialize.sh
Comment (Rosspet) [inline featb/src/initialize.sh:1]: Is this meant to have code for the initialise command?

File: infrastructure/.gitignore [hunk 1]
```diff
@@ -10,3 +10,15 @@ __pycache__
...
 provena/keycloak/customised-realm.json
+
+# configuration - keycloak themes
+provena/keycloak/themes/*
+!provena/keycloak/themes/default/*
+!provena/keycloak/themes/default
+
+# CDK configs
+configs/*
+!configs/sample*.json
```
Comment (Rosspet) [inline infrastructure/.gitignore:21]: cool nice

File: infrastructure/app_run.py [hunk 2]
```diff
@@ -55,46 +46,34 @@ def run(
...
+        # cast
+        config = cast(ProvenaUIOnlyConfig, config)
+
+        # determine stack IDs and command
+        print("UI only app config loaded and validated. Success.")
+        pipeline_stack_id = config.pipeline_stack_id
+        deployment_stack_id = config.deployment_stack_id
+        input_output_subcommand = config.cdk_input_output_subcommand
+    else: 
+        raise ValueError(f"Unexpected DeploymentType in base config: {config.type}. Report to system administrator.")
```
Comment (Rosspet) [inline infrastructure/app_run.py:76]: looks good

File: infrastructure/configs/template_custom_deployment.py
Comment (Rosspet) [inline infrastructure/configs/template_custom_deployment.py:1]: adios

File: infrastructure/provena/config/config_class.py [hunk 3]
```diff
@@ -406,17 +393,24 @@ class DNSConfig(HZConfig):
...
 
+    # config source repo that defined this config
+    config_source_repo_clone_string: str
+    # The name space for the config
+    config_source_name_space: str
+    # The stage for the config
+    config_source_stage: str
+    # The secret ARN which contains the username/token combination to use for cloning the repo
+    config_repo_oauth_token_secret_arn: str
+
```
Comment (Rosspet) [inline infrastructure/provena/config/config_class.py:413]: 💯

File: infrastructure/provena/config/config_helpers.py [hunk 4]
```diff
@@ -164,34 +164,42 @@ def subdomain_to_full(subdomain: str) -> str:
...
+        else "",
+        auth_api=subdomain_to_full(config.components.auth_api.api_domain)
+        if config.components.auth_api
+        else "",
+        warmer_api=subdomain_to_full(config.components.warmer.domain)
+        if config.components.warmer
+        else "",
+        async_jobs_api=subdomain_to_full(config.components.async_jobs.job_api_domain)
+        if config.components.async_jobs
+        else "",
```
Comment (Rosspet) [inline infrastructure/provena/config/config_helpers.py:188]: This is a bit weird, but i guess the attribute for domain is diffferent for different components so cant really be have a generic function for all these if else statements.

File: infrastructure/provena/config/config_loader.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        file_name = f"{CONFIGS_FILE_PREFIX}/{config_id.lower()}.json"
+        try:
+            with open(file_name, "r") as f:
+                content = f.read()
+        except Exception:
+            prev = file_name
+            file_name = f"{CONFIGS_FILE_PREFIX}/{config_id.upper()}.json"
+            print(f"Couldn't find {prev} trying {file_name}")
+            with open(file_name, "r") as f:
+                content = f.read()
```
Comment (Rosspet) [inline infrastructure/provena/config/config_loader.py:30]: doc strings please

File: scripts/testing/mypy_type_checks.sh [hunk 6]
```diff
@@ -2,88 +2,87 @@
...
-    "dev_requirements.txt"
-    # Shared interfaces
-    "testing-requirements.txt"
-)
-
-# List of type checking locations
-mypy_check_locations=(
+# Array of locations and requirement files
+# Each odd index is a location, and the following even index is its corresponding requirement file
+locations_and_requirements=(
```
Comment (Rosspet) [inline scripts/testing/mypy_type_checks.sh:50]: nice

File: utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    The function supports the following patterns (in order of precedence):
+    1. "${VARIABLE_NAME?true_value||false_value}" - Quoted conditional substitution with else value
+    2. "${VARIABLE_NAME?true_value}" - Quoted conditional substitution without else value
+    3. ${VARIABLE_NAME?true_value||false_value} - Unquoted conditional substitution with else value
+    4. ${VARIABLE_NAME?true_value} - Unquoted conditional substitution without else value
+    5. ${VARIABLE_NAME:-default_value} - Substitution with default value
+    6. ${VARIABLE_NAME} - Basic substitution
+
+    Args:
+        input_string (str): The input string containing variables to be substituted.
```
Comment (Rosspet) [inline utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:31]: is this input_string, the string of one of the json config files?

File: utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                    return value
+        return None
+
+    def is_truthy(value: Optional[str]) -> bool:
+        """Determine if a value is truthy."""
+        return value is not None and value.lower() != "false"
+
+    def replace_quotes(s: str) -> str:
+        """Replace single quotes with double quotes, except for 'null'."""
+        return "null" if s.strip() == "null" else s.replace("'", '"')
```
Comment (Rosspet) [inline utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:56]: What about apostrophes? E.g. Don't -> Don"t. Is that what we want? will input to this function have apostrophes? Need to also check for Null or NULL?

File: featb/src/initialize.sh
Comment (PeterBaker0) [reply featb/src/initialize.sh:1]: Nope - just placeholder

File: utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py (hunk 7 above)
Comment (PeterBaker0) [reply utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:31]: could be any config file yep

Comment (PeterBaker0) [reply utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:31]: will update

File: utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py (hunk 8 above)
Comment (PeterBaker0) [reply utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:56]: The doc string is out of date

Comment (PeterBaker0) [reply utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:56]: this is actually incorrect from what I wanted it to do - I need to update this

Comment (Rosspet) [reply utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py:56]: changing to simply return s.replace("'", '"')

**Review by Rosspet — APPROVED** [review]
Looks good - just confirm my small changes in the last commit then ready to merge. 

Looking forward to this one!! Cheers.

**Pull request merged**